File: python/proxy/proxy.py
from abc import ABC


class IPhone(ABC):
    def call(self):
        raise NotImplementedError()


class Phone(IPhone):
    def call(self):
        print("Calling!")


class ProxyPhone(IPhone):
    def __init__(self):
        self._phone = None

    def call(self):
        if self._phone is None:
            print("Creating phone!")
            self._phone = Phone()
        self._phone.call()


def main():
    phone = ProxyPhone()
    phone.call()
    phone.call()


if __name__ == '__main__':
    main()

File: cpp/proxy/proxy.cpp
#include <iostream>
#include <memory>

class IPhone {
public:
    virtual void call() = 0;
};

class Phone : public IPhone {
public:
    void call() override {
        std::cout << "Calling!" << std::endl;
    }
};

class PhoneProxy : public IPhone {
public:
    void call() override {
        if (_phone == nullptr) {
            std::cout << "Creating phone!" << std::endl;
            _phone = std::make_shared<Phone>();
        }
        _phone->call();
    }

private:
    std::shared_ptr<IPhone> _phone;
};

int main() {
    PhoneProxy phoneProxy;
    phoneProxy.call();
    phoneProxy.call();
    return 0;
}
